File: workspace/core/server/src/models/workflow/prompts.rs
use async_graphql::*;
use serde_json::Value;
use tokio_postgres::Row;
use uuid::Uuid;

pub struct Prompt {
    pub id: Uuid,
    pub name: String,
    pub description: String,
    pub system_prompt: String,
    pub user_prompt: String,
    pub input_type: String,
    pub output_type: String,
    pub schema: Option<Value>
}

#[derive(InputObject)]
pub struct PromptInput {
    pub name: String,
    pub description: String,
    pub system_prompt: String,
    pub user_prompt: String,
    pub input_type: String,
    pub output_type: String,
    pub schema: Option<Value>,
}

impl From<&Row> for Prompt {
    fn from(row: &Row) -> Self {
        Self {
            id: row.get("id"),
            name: row.get("name"),
            description: row.get("description"),
            system_prompt: row.get("system_prompt"),
            user_prompt: row.get("user_prompt"),
            input_type: row.get("input_type"),
            output_type: row.get("output_type"),
            schema: row.get("schema"),
        }
    }
}

File: workspace/runners/kotlin/installers/src/main/kotlin/io/bosca/workflow/ext/Prompts.kt
package io.bosca.workflow.ext

import io.bosca.graphql.type.PromptInput
import io.bosca.util.encodeToOptional
import io.bosca.util.toOptional
import io.bosca.workflow.models.PromptDefinition

fun PromptDefinition.toInput() = PromptInput(
    name = name,
    description = description,
    inputType = inputType,
    outputType = outputType,
    systemPrompt = systemPrompt,
    userPrompt = userPrompt,
    schema = schema?.encodeToOptional() ?: emptyMap<String, Any>().toOptional()
)